docs/en/tutorials/create-wax-rng-smart-contract/rng_sample.md

```diff
@@ -1,141 +1,230 @@
+      if(iSigningValue != orng::signvals.end()){
+         signing_value <<= 1;
+         c++;
+      }
+   }
+
+   // Ok, I don't think this will happen but... what if it does?
+   check(c < 64, "No combination was valid? Inconceivable!");
+
+   // Prepare the table where the random value will be received.
+   _customers.emplace(get_self(), [&](auto& rec) {
+      rec.signing_value = signing_value;
+      rec.customer_id = customer_id;
+   });
+
+   // Call to orng.wax smart contract
+   action(
+      { get_self(), "active"_n },
+      "orng.wax"_n,
+      "requestrand"_n,
+      std::tuple{ signing_value, signing_value, get_self() })
+      .send();
+}
+```
+
+*transaction_id* is a unique number but since we have to select only a 64-bit portion it is possible that the resulting value is not unique. For that reason we perform the check against the `signvals.a` table and, if the result is a number already chosen, we perform a 1-bit rotation to try again.
+
+When we are certain that we have found a unique number, we store the number, along with the user's name, and call WAX RNG by sending it the seed (*signing_value*) and the same value as the call identifier. We could use the user's own name as the identifier (*customer_id.value*) but this would complicate the code in case a user requests several random numbers and his requests are simultaneously stored in the table `rngcustomers` whose primary key is *signing_value*.
+
+4.2. ACTION receiverand
+
+Callback action that will be called from WAX RNG to return us, in case of success, the random number, together with the code that we provide as identifier of the request (*signing_value*).
+
+**Parameters**
+
+| Parameter   | Type | Description                                       |
+| ----------- | ---- | ---------------------------------------------------- |
+| signing_value | uint64_t    | Code sent as identifier.                                   |
+| random_value  | checksum256 | Hash code generated by WAX RNG that we will use to obtain the random number |
+
+```c++
+ACTION rngtest::receiverand(uint64_t signing_value, const checksum256& random_value) {
+   require_auth(ORNG_CONTRACT);
+
+   //cast the random_value to a smaller number
+   uint64_t max_value = 100;
+   auto byte_array = random_value.extract_as_byte_array();
+
+   uint64_t random_int = 0;
+   for (int i = 0; i < 8; i++) {
+      random_int <<= 8;
+      random_int |= (uint64_t)byte_array[i];
+   }
+
+   uint64_t num1 = random_int % max_value;
+
+   auto iCustomers = _customers.require_find(signing_value, "Error: Petition not found!");
+
+   // update table with random_value
+   _customers.modify(iCustomers, get_self(), [&](auto& rec) {
+      rec.random_value = random_value;
+      rec.final_number = num1;
+   });
+}
+```
+We make sure that only WAX RNG will be able to call this action with *require_auth*.
+
+We extract the first 8 bits of the returned random number and use it to get a number no larger than 100. 
```

Comment:
This isn't correct; they are just rebuilding a 64-bit uint 8 bits at a type (shift-left the sum by 8, then add the next byte in the index.

This is the same as 

`(byte_array[0] << 56) + (byte_array[1] << 48) + (byte_array[2] << 40) + (byte_array[3] << 32) + (byte_array[4] << 24) + (byte_array[5] << 16) + (byte_array[6] << 8) + byte_array[7];`

(just Imagine all the bytes are cast to uint64_t before shifting; left that out to reduce noise)

Reply:
Yep. Sometimes we make the easy difficult and then we get lost.
This is the correction:
```c++
auto byte_array = random_value.extract_as_byte_array();
uint8_t random_int = 0;
random_int = byte_array[0];
uint64_t num1 = random_int % max_value;
```

Reply:
🙏